File: crates/fortress-go/tenant.go
package fortress

import (
	"fmt"
	"sync"
	"time"
)

// TenantManager manages tenant isolation and configuration
type TenantManager struct {
	config  *Config
	tenants map[string]*TenantConfig
	stats   map[string]*TenantStats
	mu      sync.RWMutex
}

// NewTenantManager creates a new tenant manager
func NewTenantManager(config *Config) *TenantManager {
	return &TenantManager{
		config:  config,
		tenants: make(map[string]*TenantConfig),
		stats:   make(map[string]*TenantStats),
	}
}

// CreateTenant creates a new tenant
func (t *TenantManager) CreateTenant(config *TenantConfig) (string, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	tenantID := GenerateRandomString(16)
	tenantConfig := *config
	tenantConfig.ID = tenantID

	t.tenants[tenantID] = &tenantConfig
	t.stats[tenantID] = t.createDefaultStats(tenantID)

	return tenantID, nil
}

// GetTenant retrieves tenant configuration by ID
func (t *TenantManager) GetTenant(tenantID string) *TenantConfig {
	t.mu.RLock()
	defer t.mu.RUnlock()

	tenant, exists := t.tenants[tenantID]
	if !exists {
		return nil
	}
	return tenant
}

// UpdateTenant updates tenant configuration
func (t *TenantManager) UpdateTenant(tenantID string, updates *TenantConfig) (bool, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	tenant, exists := t.tenants[tenantID]
	if !exists {
		return false, NewTenantError("tenant not found", nil)
	}

	// Update fields
	if updates.Name != "" {
		tenant.Name = updates.Name
	}
	if updates.Description != "" {
		tenant.Description = updates.Description
	}
	if updates.IsolationLevel != "" {
		tenant.IsolationLevel = updates.IsolationLevel
	}

	return true, nil
}

// DeleteTenant deletes a tenant
func (t *TenantManager) DeleteTenant(tenantID string) (bool, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	_, exists := t.tenants[tenantID]
	if exists {
		delete(t.tenants, tenantID)
		delete(t.stats, tenantID)
		return true, nil
	}
	return false, nil
}

// ListTenants returns all tenant IDs
func (t *TenantManager) ListTenants() []string {
	t.mu.RLock()
	defer t.mu.RUnlock()

	var tenantIDs []string
	for tenantID := range t.tenants {
		tenantIDs = append(tenantIDs, tenantID)
	}
	return tenantIDs
}

// GetTenantStats retrieves tenant statistics
func (t *TenantManager) GetTenantStats(tenantID string) *TenantStats {
	t.mu.RLock()
	defer t.mu.RUnlock()

	stats, exists := t.stats[tenantID]
	if !exists {
		return nil
	}
	return stats
}

// UpdateStats updates tenant statistics
func (t *TenantManager) UpdateStats(tenantID string, operation string) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	stats, exists := t.stats[tenantID]
	if !exists {
		return NewTenantError("tenant not found", nil)
	}

	switch operation {
	case "encrypt":
		stats.TotalEncryptions++
	case "decrypt":
		stats.TotalDecryptions++
	}

	stats.LastActivity = time.Now()
	return nil
}

// CheckResourceLimits checks if tenant is within resource limits
func (t *TenantManager) CheckResourceLimits(tenantID string, resource string, amount int) (bool, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	tenant, exists := t.tenants[tenantID]
	stats, statsExists := t.stats[tenantID]

	if !exists || !statsExists {
		return false, NewTenantError("tenant not found", nil)
	}

	limits := tenant.ResourceLimits

	switch resource {
	case "keys":
		return stats.TotalKeys+amount <= limits.MaxKeys, nil
	case "storage":
		return int64(stats.StorageUsageBytes)+int64(amount) <= int64(limits.MaxStorageBytes), nil
	case "encryptions":
		return stats.TotalEncryptions+amount <= limits.MaxEncryptionsPerHour, nil
	case "decryptions":
		return stats.TotalDecryptions+amount <= limits.MaxDecryptionsPerHour, nil
	case "concurrent_operations":
		return amount <= limits.MaxConcurrentOperations, nil
	default:
		return true, nil
	}
}

// GetAllStats returns all tenant statistics
func (t *TenantManager) GetAllStats() map[string]*TenantStats {
	t.mu.RLock()
	defer t.mu.RUnlock()

	result := make(map[string]*TenantStats)
	for tenantID, stats := range t.stats {
		// Return a copy to avoid mutation
		statsCopy := *stats
		result[tenantID] = &statsCopy
	}
	return result
}

// ResetStats resets tenant statistics
func (t *TenantManager) ResetStats(tenantID string) (bool, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	_, exists := t.tenants[tenantID]
	if !exists {
		return false, NewTenantError("tenant not found", nil)
	}

	t.stats[tenantID] = t.createDefaultStats(tenantID)
	return true, nil
}

// ValidateTenantConfig validates tenant configuration
func (t *TenantManager) ValidateTenantConfig(config *TenantConfig) error {
	if config.ID == "" {
		return NewValidationError("tenant ID is required", nil)
	}

	if config.Name == "" {
		return NewValidationError("tenant name is required", nil)
	}

	validIsolationLevels := []string{"strict", "shared", "partial"}
	isolationValid := false
	for _, level := range validIsolationLevels {
		if config.IsolationLevel == level {
			isolationValid = true
			break
		}
	}
	if !isolationValid {
		return NewValidationError("invalid tenant isolation level", nil)
	}

	if config.ResourceLimits.MaxKeys <= 0 {
		return NewValidationError("max keys must be positive", nil)
	}

	if config.ResourceLimits.MaxStorageBytes <= 0 {
		return NewValidationError("max storage bytes must be positive", nil)
	}

	return nil
}

// GetTenantsByIsolationLevel returns tenants by isolation level
func (t *TenantManager) GetTenantsByIsolationLevel(level string) []string {
	t.mu.RLock()
	defer t.mu.RUnlock()

	var tenantIDs []string
	for tenantID, tenant := range t.tenants {
		if tenant.IsolationLevel == level {
			tenantIDs = append(tenantIDs, tenantID)
		}
	}
	return tenantIDs
}

// GetTenantsNearLimits returns tenants approaching resource limits
func (t *TenantManager) GetTenantsNearLimits(threshold float64) []TenantLimitWarning {
	t.mu.RLock()
	defer t.mu.RUnlock()

	var warnings []TenantLimitWarning

	for tenantID, tenant := range t.tenants {
		stats := t.stats[tenantID]
		limits := tenant.ResourceLimits

		// Check each resource
		checks := []struct {
			resource string
			usage    int
			limit    int
		}{
			{"keys", stats.TotalKeys, limits.MaxKeys},
			{"storage", stats.StorageUsageBytes, limits.MaxStorageBytes},
			{"encryptions", stats.TotalEncryptions, limits.MaxEncryptionsPerHour},
			{"decryptions", stats.TotalDecryptions, limits.MaxDecryptionsPerHour},
		}

		for _, check := range checks {
			if check.limit > 0 && float64(check.usage)/float64(check.limit) >= threshold {
				warnings = append(warnings, TenantLimitWarning{
					TenantID: tenantID,
					Resource: check.resource,
					Usage:    check.usage,
					Limit:    check.limit,
				})
			}
		}
	}

	return warnings
}

// createDefaultStats creates default statistics for a tenant
func (t *TenantManager) createDefaultStats(tenantID string) *TenantStats {
	return &TenantStats{
		TenantID:          tenantID,
		TotalKeys:         0,
		ActiveKeys:        0,
		TotalEncryptions:   0,
		TotalDecryptions:   0,
		StorageUsageBytes: 0,
		AuditEntries:      0,
		LastActivity:      time.Now(),
	}
}

// TenantLimitWarning represents a warning about resource limits
type TenantLimitWarning struct {
	TenantID string `json:"tenant_id"`
	Resource string `json:"resource"`
	Usage    int    `json:"usage"`
	Limit    int    `json:"limit"`
}

File: crates/fortress-js/src/tenant.ts
/**
 * Tenant management functionality for Fortress JavaScript/TypeScript SDK
 */

import { FortressConfig } from './config';
import { FortressError } from './error';
import { TenantConfig, TenantStats, ResourceLimits } from './types';

/**
 * Tenant manager class
 */
export class TenantManager {
  private config: FortressConfig;
  private tenants: Map<string, TenantConfig>;
  private stats: Map<string, TenantStats>;

  constructor(config: FortressConfig) {
    this.config = config;
    this.tenants = new Map();
    this.stats = new Map();
    
    // Initialize with default tenant
    this.tenants.set('default', config.tenantConfig);
    this.stats.set('default', this.createDefaultStats('default'));
  }

  /**
   * Create a new tenant
   */
  async createTenant(config: Omit<TenantConfig, 'id'>): Promise<string> {
    try {
      const tenantId = this.generateId();
      const tenantConfig: TenantConfig = {
        id: tenantId,
        ...config,
      };

      this.tenants.set(tenantId, tenantConfig);
      this.stats.set(tenantId, this.createDefaultStats(tenantId));

      return tenantId;
    } catch (error) {
      throw FortressError.tenant('Failed to create tenant', error as Error);
    }
  }

  /**
   * Get tenant configuration
   */
  getTenant(tenantId: string): TenantConfig | null {
    return this.tenants.get(tenantId) || null;
  }

  /**
   * Update tenant configuration
   */
  async updateTenant(tenantId: string, updates: Partial<TenantConfig>): Promise<boolean> {
    try {
      const tenant = this.tenants.get(tenantId);
      if (!tenant) {
        return false;
      }

      const updatedTenant = { ...tenant, ...updates };
      this.tenants.set(tenantId, updatedTenant);

      return true;
    } catch (error) {
      throw FortressError.tenant('Failed to update tenant', error as Error);
    }
  }

  /**
   * Delete a tenant
   */
  async deleteTenant(tenantId: string): Promise<boolean> {
    try {
      const deleted = this.tenants.delete(tenantId);
      if (deleted) {
        this.stats.delete(tenantId);
      }
      return deleted;
    } catch (error) {
      throw FortressError.tenant('Failed to delete tenant', error as Error);
    }
  }

  /**
   * List all tenants
   */
  listTenants(): string[] {
    return Array.from(this.tenants.keys());
  }

  /**
   * Get tenant statistics
   */
  getTenantStats(tenantId: string): TenantStats | null {
    return this.stats.get(tenantId) || null;
  }

  /**
   * Update tenant statistics
   */
  async updateStats(tenantId: string, operation: 'encrypt' | 'decrypt'): Promise<void> {
    try {
      const stats = this.stats.get(tenantId);
      if (!stats) {
        return;
      }

      if (operation === 'encrypt') {
        stats.total_encryptions++;
      } else {
        stats.total_decryptions++;
      }

      stats.last_activity = new Date();
    } catch (error) {
      throw FortressError.tenant('Failed to update stats', error as Error);
    }
  }

  /**
   * Check resource limits
   */
  async checkResourceLimits(tenantId: string, resource: string, amount: number = 1): Promise<boolean> {
    try {
      const tenant = this.tenants.get(tenantId);
      const stats = this.stats.get(tenantId);
      
      if (!tenant || !stats) {
        return false;
      }

      const limits = tenant.resource_limits;

      switch (resource) {
        case 'keys':
          return stats.total_keys + amount <= limits.max_keys;
        case 'storage':
          return stats.storage_usage_bytes + amount <= limits.max_storage_bytes;
        case 'encryptions':
          return stats.total_encryptions + amount <= limits.max_encryptions_per_hour;
        case 'decryptions':
          return stats.total_decryptions + amount <= limits.max_decryptions_per_hour;
        case 'concurrent_operations':
          return amount <= limits.max_concurrent_operations;
        default:
          return true;
      }
    } catch (error) {
      throw FortressError.tenant('Failed to check resource limits', error as Error);
    }
  }

  /**
   * Get all tenant statistics
   */
  getAllStats(): Map<string, TenantStats> {
    return new Map(this.stats);
  }

  /**
   * Reset tenant statistics
   */
  async resetStats(tenantId: string): Promise<boolean> {
    try {
      const tenant = this.tenants.get(tenantId);
      if (!tenant) {
        return false;
      }

      this.stats.set(tenantId, this.createDefaultStats(tenantId));
      return true;
    } catch (error) {
      throw FortressError.tenant('Failed to reset stats', error as Error);
    }
  }

  /**
   * Validate tenant configuration
   */
  validateTenantConfig(config: TenantConfig): boolean {
    try {
      // Check required fields
      if (!config.id || !config.name || !config.isolation_level) {
        return false;
      }

      // Check resource limits
      const limits = config.resource_limits;
      if (limits.max_keys <= 0 || 
          limits.max_storage_bytes <= 0 ||
          limits.max_encryptions_per_hour <= 0 ||
          limits.max_decryptions_per_hour <= 0 ||
          limits.max_concurrent_operations <= 0) {
        return false;
      }

      // Check isolation level
      const validIsolationLevels = ['strict', 'shared', 'partial'];
      if (!validIsolationLevels.includes(config.isolation_level)) {
        return false;
      }

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Create default statistics for a tenant
   */
  private createDefaultStats(tenantId: string): TenantStats {
    return {
      tenant_id: tenantId,
      total_keys: 0,
      active_keys: 0,
      total_encryptions: 0,
      total_decryptions: 0,
      storage_usage_bytes: 0,
      audit_entries: 0,
      last_activity: new Date(),
    };
  }

  /**
   * Generate unique ID for tenants
   */
  private generateId(): string {
    return 'tenant_' + Math.random().toString(36).substr(2, 9) + Date.now().toString(36);
  }

  /**
   * Get tenant by isolation level
   */
  getTenantsByIsolationLevel(level: 'strict' | 'shared' | 'partial'): string[] {
    return Array.from(this.tenants.values())
      .filter(tenant => tenant.isolation_level === level)
      .map(tenant => tenant.id);
  }

  /**
   * Get tenants approaching resource limits
   */
  getTenantsNearLimits(threshold: number = 0.8): Array<{ tenantId: string; resource: string; usage: number; limit: number }> {
    const results: Array<{ tenantId: string; resource: string; usage: number; limit: number }> = [];

    for (const [tenantId, tenant] of this.tenants) {
      const stats = this.stats.get(tenantId);
      if (!stats) continue;

      const limits = tenant.resource_limits;

      // Check each resource
      const checks = [
        { resource: 'keys', usage: stats.total_keys, limit: limits.max_keys },
        { resource: 'storage', usage: stats.storage_usage_bytes, limit: limits.max_storage_bytes },
        { resource: 'encryptions', usage: stats.total_encryptions, limit: limits.max_encryptions_per_hour },
        { resource: 'decryptions', usage: stats.total_decryptions, limit: limits.max_decryptions_per_hour },
      ];

      for (const check of checks) {
        if (check.limit > 0 && (check.usage / check.limit) >= threshold) {
          results.push({
            tenantId,
            ...check,
          });
        }
      }
    }

    return results;
  }
}
